database/src/database.rs

```diff
@@ -65,4 +133,35 @@ pub trait DatabaseRw<T: Table>: DatabaseRo<T> {
     /// # Errors
     /// This will return [`RuntimeError::KeyNotFound`] wrapped in [`Err`] if `key` does not exist.
     fn delete(&mut self, key: &T::Key) -> Result<(), RuntimeError>;
+
+    /// TODO
+    ///
+    /// # Errors
+    /// TODO
+    fn clear(&mut self) -> Result<(), RuntimeError>;
+
+    /// TODO
+    ///
+    /// - `true == keep`
+    /// - `false == remove`
+    ///
+    /// # Errors
+    /// TODO
+    ///
+    /// TODO: Document early return on error.
+    fn retain<P>(&mut self, predicate: P) -> Result<(), RuntimeError>
+    where
+        P: FnMut(T::Key, T::Value) -> bool;
```

Comment:
What should be the behavior on database errors mid-`retain()`?

Currently it just early returns, so all removes up until that point are still `commit()`-able. I think this makes the most sense since you can `abort()` if there's an error?